src/classes/frDonation.cls

```diff
@@ -53,84 +55,126 @@ public class frDonation extends frModel implements frSyncable {
             } else {
                 frUtil.logRelationshipError(getFrType(), funraiseId, 
                                             frUtil.Entity.SUBSCRIPTION, subscriptionId);
             }
         }
-        Opportunity opp = this.o;
-        
-        
+
+        //connect pledge
         Boolean isPledge = Boolean.valueOf(request.get('pledge'));
-        Boolean alreadyMappedToPledge = [SELECT COUNT() FROM Opportunity 
-                                         WHERE fr_Id__c = :funraiseId AND Funraise_Pledge__c != null] > 0;
-        if(!isPledge && !alreadyMappedToPledge) {
-            Pledge__c pledge = frPledge.findActive(opp.fr_Donor__c, opp.CloseDate);
-            opp.Funraise_Pledge__c = pledge != null ? pledge.Id : null;
+        if(frUtil.hasNPCobjects()) {
+            List<SObject> alreadyMappedToPledgeList= Database.query('SELECT id FROM GiftTransaction WHERE fr_Id__c = :funraiseId AND Funraise_Pledge__c != null');
+            Boolean alreadyMappedToPledge = alreadyMappedToPledgeList.size() > 0;
+            if(!isPledge && !alreadyMappedToPledge) {
+                Pledge__c pledge = frPledge.findActiveAcc((Id) o.get('DonorId'), (Date) o.get('CheckDate'));
+                NPC_Setting__c npcSetting = NPC_Setting__c.getOrgDefaults();
+                if (npcSetting != null && pledge != null) {
+                    String pledgeId = npcSetting.NPCFieldName__c;
+                    o.put('Funraise_GT_Pledge__c', pledge.id);
+                } else {
+                    System.debug('NPC Setting not found at the organization level.');
+                }
+            }
         }
-        
+        else {
+            Opportunity opp = (Opportunity) this.o;
+            Boolean alreadyMappedToPledge = [SELECT COUNT() FROM Opportunity 
+                                            WHERE fr_Id__c = :funraiseId AND Funraise_Pledge__c != null] > 0;
+            if(!isPledge && !alreadyMappedToPledge) {
+                Pledge__c pledge = frPledge.findActive(opp.fr_Donor__c, opp.CloseDate);
+                opp.Funraise_Pledge__c = pledge != null ? pledge.Id : null;
+            }
+        }
+
+        //upsert
         try {
-            Database.upsert(opp, Opportunity.Fields.fr_Id__c, true);
+            Database.upsert(o, true);
             result = true;
         } catch (Exception ex) {
             frUtil.logException(getFrType(), funraiseId, ex);
         }
-        
-        if(isPledge && opp.Id != null) {
-            frPledge.create(opp);
+
+        if(isPledge && o.get('Id') != null) {
+            frPledge.create(o);
         }
         
-        if(opp.Id != null) {
+        if(o.get('Id') != null && !frUtil.hasNPCobjects()) {
+            createOpportunityMapping(request);
+        }
+
+        if(Test.isRunningTest() && String.valueOf(request.get('APEX_TESTS')) == 'true' && frUtil.hasNPCobjects()) {
+            setOpportunityStage(new Opportunity(), 'Closed Won');
```

Comment:
Can you explain why we are setting the stage on a new Opportunity here instead of the object we are working with `o` ?

Reply:
This is for testing purposes! since in an NPC org this obj would not be an opportunity but we still need to test it for coverage